R/generate_subsample.R

```diff
@@ -0,0 +1,79 @@
+#' Generate Subsample
+#' @description Subsamples from sampling events in a season.
+#' @details This function subsamples from all samples in a given season according to the following logic:
+#'
+#' * At least 50% of samples per site per event will be sampled.
+#' * If the number of samples is odd, divide that number by 2 and round the resulting number up to the nearest integer.
+#' * If the total number of samples for a given site in a given event is less than 20, process all samples for that site/event.
+#' * If multiple bins are represented in a set of samples for a given site and event, select 50% of the samples from each bin for processing.
+#' * If the total number of samples in a bin is less than or equal to 5, process all of the samples for that bin.
+#' * If this rule contradicts the “less than 20” rule (above), this rule should be prioritized. For example, if we receive a sample set from a given site and event where Bins A, B, C, D, and E are each represented by five samples (total sample size = 25), process all of the samples for that site/event.
+#' * Subsampling should be random.
+#' @param con connection to the database
+#' @param season The season for which you want to pull subsamples. A season consists of all sampling events
+#' from the given year up to September 30th and from the previous year after October 1st.
+#' @examples
+#' # connect with database
+#' con <- gr_db_connect()
+#'
+#' # subsample from all sampling events in season 2024
+#' subsamples <- generate_subsample(con, 2024)
+#' @returns A table containing the following columns:
+#'
+#' * **sample_id**
+#' * **datetime_collected**
+#' * **stream_name**
+#' * **sample_bin_code**
+#' * **sample_event_number**
+#' @export
+#' @md
+generate_subsample <- function(con, season) {
+  if(!is.numeric(season)) {
+    cli::cli_abort("Season must be a numeric value.")
+  }
+  if(nchar(season) != 4) {
+    cli::cli_abort("Season must be in the format YYYY (i.e. 2024).")
+  }
+  # get all sample event IDs for a season
+  samples <- grunID::filter_dataset(con, season) |>
```

Comment:
now that `grunID::filter_dataset` is being used in more places in the code base we should rename it to something more specific like `sample_filter_to_season` or something like that, just to make it clear what it returns